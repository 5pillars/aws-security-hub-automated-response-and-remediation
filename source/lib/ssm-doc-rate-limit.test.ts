// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { CfnDocument } from 'aws-cdk-lib/aws-ssm';
import { Aspects, Stack } from 'aws-cdk-lib';
import SsmDocRateLimit from './ssm-doc-rate-limit';
import { WaitProvider } from './wait-provider';
import { Template } from 'aws-cdk-lib/assertions';

describe('SSM doc rate limit aspect', function () {
  it('configures dependencies for single document', function () {
    const stack = new Stack();
    const serviceToken = 'my-token';
    const waitProvider = WaitProvider.fromServiceToken(stack, 'WaitProvider', serviceToken);
    Aspects.of(stack).add(new SsmDocRateLimit(waitProvider));
    const content = {};
    new CfnDocument(stack, 'Document', { content });
    const template = Template.fromStack(stack);

    const documents = template.findResources('AWS::SSM::Document', { Properties: { Content: content } });
    const documentLogicalIds = Object.getOwnPropertyNames(documents);
    expect(documentLogicalIds).toHaveLength(1);
    const documentLogicalId = documentLogicalIds[0];
    const document = documents[documentLogicalId];

    const createWaits = template.findResources('Custom::Wait', {
      Properties: {
        CreateIntervalSeconds: 1,
        UpdateIntervalSeconds: 1,
        DeleteIntervalSeconds: 0,
        ServiceToken: serviceToken,
      },
    });
    const createWaitLogicalIds = Object.getOwnPropertyNames(createWaits);
    expect(createWaitLogicalIds).toHaveLength(1);
    const createWaitLogicalId = createWaitLogicalIds[0];
    expect(document.DependsOn).toEqual(expect.arrayContaining([createWaitLogicalId]));

    const deleteWaits = template.findResources('Custom::Wait', {
      Properties: {
        CreateIntervalSeconds: 0,
        UpdateIntervalSeconds: 0,
        DeleteIntervalSeconds: 0.5,
        ServiceToken: serviceToken,
      },
    });
    const deleteWaitLogicalIds = Object.getOwnPropertyNames(deleteWaits);
    expect(deleteWaitLogicalIds).toHaveLength(1);
    const deleteWait = deleteWaits[deleteWaitLogicalIds[0]];
    expect(deleteWait.DependsOn).toEqual(expect.arrayContaining([documentLogicalId]));
  });

  it('configures dependencies for many documents', function () {
    const stack = new Stack();
    const serviceToken = 'my-token';
    const waitProvider = WaitProvider.fromServiceToken(stack, 'WaitProvider', serviceToken);
    Aspects.of(stack).add(new SsmDocRateLimit(waitProvider));
    const content = {};
    const numDocuments = 12;
    for (let i = 0; i < numDocuments; ++i) {
      new CfnDocument(stack, `Document${i}`, { content });
    }
    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();

    const documents = template.findResources('AWS::SSM::Document', { Properties: { Content: content } });
    const documentLogicalIds = Object.getOwnPropertyNames(documents);
    expect(documentLogicalIds).toHaveLength(numDocuments);

    const expectedBatchSize = 5;

    const createWaits = template.findResources('Custom::Wait', {
      Properties: {
        CreateIntervalSeconds: 1,
        UpdateIntervalSeconds: 1,
        DeleteIntervalSeconds: 0,
        ServiceToken: serviceToken,
      },
    });
    const createWaitLogicalIds = Object.getOwnPropertyNames(createWaits);
    expect(createWaitLogicalIds).toHaveLength(Math.ceil(numDocuments / expectedBatchSize));

    const deleteWaits = template.findResources('Custom::Wait', {
      Properties: {
        CreateIntervalSeconds: 0,
        UpdateIntervalSeconds: 0,
        DeleteIntervalSeconds: 0.5,
        ServiceToken: serviceToken,
      },
    });
    const deleteWaitLogicalIds = Object.getOwnPropertyNames(deleteWaits);
    expect(deleteWaitLogicalIds).toHaveLength(Math.ceil(numDocuments / expectedBatchSize));
  });
});
