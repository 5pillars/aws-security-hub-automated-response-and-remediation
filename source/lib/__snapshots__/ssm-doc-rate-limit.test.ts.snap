// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSM doc rate limit aspect configures dependencies for many documents 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CreateWait0": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "CreateIntervalSeconds": 1,
        "DeleteIntervalSeconds": 0,
        "DocumentPropertiesHash": "97f8546cc30d8ca122f75268f86704de82b2760257dd324937478bfa046400cf",
        "ServiceToken": "my-token",
        "UpdateIntervalSeconds": 1,
      },
      "Type": "Custom::Wait",
      "UpdateReplacePolicy": "Delete",
    },
    "CreateWait1": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "CreateWait0",
      ],
      "Properties": {
        "CreateIntervalSeconds": 1,
        "DeleteIntervalSeconds": 0,
        "DocumentPropertiesHash": "97f8546cc30d8ca122f75268f86704de82b2760257dd324937478bfa046400cf",
        "ServiceToken": "my-token",
        "UpdateIntervalSeconds": 1,
      },
      "Type": "Custom::Wait",
      "UpdateReplacePolicy": "Delete",
    },
    "CreateWait2": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "CreateWait1",
      ],
      "Properties": {
        "CreateIntervalSeconds": 1,
        "DeleteIntervalSeconds": 0,
        "DocumentPropertiesHash": "3758b126bf0cb781dcfd65289a956ffe11d35b5948f020cef3efad3937173f1d",
        "ServiceToken": "my-token",
        "UpdateIntervalSeconds": 1,
      },
      "Type": "Custom::Wait",
      "UpdateReplacePolicy": "Delete",
    },
    "DeletWait0": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "Document0",
        "Document1",
        "Document2",
        "Document3",
        "Document4",
      ],
      "Properties": {
        "CreateIntervalSeconds": 0,
        "DeleteIntervalSeconds": 0.5,
        "DocumentPropertiesHash": "97f8546cc30d8ca122f75268f86704de82b2760257dd324937478bfa046400cf",
        "ServiceToken": "my-token",
        "UpdateIntervalSeconds": 0,
      },
      "Type": "Custom::Wait",
      "UpdateReplacePolicy": "Delete",
    },
    "DeletWait1": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "DeletWait0",
        "Document5",
        "Document6",
        "Document7",
        "Document8",
        "Document9",
      ],
      "Properties": {
        "CreateIntervalSeconds": 0,
        "DeleteIntervalSeconds": 0.5,
        "DocumentPropertiesHash": "97f8546cc30d8ca122f75268f86704de82b2760257dd324937478bfa046400cf",
        "ServiceToken": "my-token",
        "UpdateIntervalSeconds": 0,
      },
      "Type": "Custom::Wait",
      "UpdateReplacePolicy": "Delete",
    },
    "DeletWait2": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "DeletWait1",
        "Document10",
        "Document11",
      ],
      "Properties": {
        "CreateIntervalSeconds": 0,
        "DeleteIntervalSeconds": 0.5,
        "DocumentPropertiesHash": "3758b126bf0cb781dcfd65289a956ffe11d35b5948f020cef3efad3937173f1d",
        "ServiceToken": "my-token",
        "UpdateIntervalSeconds": 0,
      },
      "Type": "Custom::Wait",
      "UpdateReplacePolicy": "Delete",
    },
    "Document0": {
      "DependsOn": [
        "CreateWait0",
      ],
      "Properties": {
        "Content": {},
      },
      "Type": "AWS::SSM::Document",
    },
    "Document1": {
      "DependsOn": [
        "CreateWait0",
      ],
      "Properties": {
        "Content": {},
      },
      "Type": "AWS::SSM::Document",
    },
    "Document10": {
      "DependsOn": [
        "CreateWait2",
      ],
      "Properties": {
        "Content": {},
      },
      "Type": "AWS::SSM::Document",
    },
    "Document11": {
      "DependsOn": [
        "CreateWait2",
      ],
      "Properties": {
        "Content": {},
      },
      "Type": "AWS::SSM::Document",
    },
    "Document2": {
      "DependsOn": [
        "CreateWait0",
      ],
      "Properties": {
        "Content": {},
      },
      "Type": "AWS::SSM::Document",
    },
    "Document3": {
      "DependsOn": [
        "CreateWait0",
      ],
      "Properties": {
        "Content": {},
      },
      "Type": "AWS::SSM::Document",
    },
    "Document4": {
      "DependsOn": [
        "CreateWait0",
      ],
      "Properties": {
        "Content": {},
      },
      "Type": "AWS::SSM::Document",
    },
    "Document5": {
      "DependsOn": [
        "CreateWait1",
      ],
      "Properties": {
        "Content": {},
      },
      "Type": "AWS::SSM::Document",
    },
    "Document6": {
      "DependsOn": [
        "CreateWait1",
      ],
      "Properties": {
        "Content": {},
      },
      "Type": "AWS::SSM::Document",
    },
    "Document7": {
      "DependsOn": [
        "CreateWait1",
      ],
      "Properties": {
        "Content": {},
      },
      "Type": "AWS::SSM::Document",
    },
    "Document8": {
      "DependsOn": [
        "CreateWait1",
      ],
      "Properties": {
        "Content": {},
      },
      "Type": "AWS::SSM::Document",
    },
    "Document9": {
      "DependsOn": [
        "CreateWait1",
      ],
      "Properties": {
        "Content": {},
      },
      "Type": "AWS::SSM::Document",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
